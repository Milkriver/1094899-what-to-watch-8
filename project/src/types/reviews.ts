export interface IReview {
    id: number,
    movieId: number,
    author: string,
    date: string,
    rating: number,
    text: string,
  }
