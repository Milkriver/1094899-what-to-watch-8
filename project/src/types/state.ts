import { AuthorizationStatus } from '../const';
import { IMovie } from './common';
import { IReviewResponse } from './reviews';

export interface IState {
  currentGenre: string
  movies: IMovie[]
  isDataLoaded: boolean
  authorizationStatus: AuthorizationStatus,
  movie: IMovie,
  sameMovies: IMovie[],
  reviews: IReviewResponse[],
  promoMovie: IMovie,
}
