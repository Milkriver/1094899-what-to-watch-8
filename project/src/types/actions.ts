import { AxiosInstance } from 'axios';
import { Action } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';
import { IState } from './state';

export enum ActionType {
  ChangeGenre = 'movieList/ChangeGenre',
  LoadMovies = 'movieList/LoadMovies',
  LoadSingleMovie = 'movieCard/LoadSingleMovie',
  LoadActiveMovie = 'movieCard/LoadActiveMovie',
  LoadSameGenreMovies = 'movieCard/LoadSameGenreMovies',
  LoadPromoMovie = 'movieCard/LoadPromoMovie',
  LoadFavouriteMovies = 'userList/LoadFavouriteMovies',
  AddFavoriteMovie = 'userList/AddFavoriteMovie',
  RemoveFavoriteMovie = 'userList/RemoveFavoriteMovie',
  LoadReviews = 'movieCard/LoadReviews',
  AddReview = 'movieCard/AddReview',
  RequireAuthorization = 'user/requireAuthorization',
  RequireLogout = 'user/requireLogout',
  LoadUserData = 'user/loadUserData',
}

export type ThunkActionResult<R = Promise<void>> = ThunkAction<R, IState, AxiosInstance, Action>;

export type ThunkAppDispatch = ThunkDispatch<IState, AxiosInstance, Action>;
