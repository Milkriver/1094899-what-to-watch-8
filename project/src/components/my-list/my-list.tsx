import { MovieCardElement } from './../movie-card-element/movie-card-element';
import { IMovie } from '../../types/common';
import { useEffect, useState } from 'react';
import { HeaderLogo } from '../header-logo/header-logo';
import HeaderUserBlock from '../header-user-block/header-user-block';
import { Footer } from '../footer/footer';
import { IState } from '../../types/state';
import { ThunkAppDispatch } from '../../types/actions';
import { fetchFavoriteMovies } from '../../store/api-actions';
import { connect, ConnectedProps } from 'react-redux';


const mapStateToProps = ({ favouriteMovies }: IState) => ({
  favouriteMovies,
});

const mapDispatchToProps = (dispatch: ThunkAppDispatch) => ({
  onFetchFavouriteMovies() {
    dispatch(fetchFavoriteMovies());
  },
});

const connector = connect(mapStateToProps, mapDispatchToProps);
type PropsFromRedux = ConnectedProps<typeof connector>;

function MyList({ favouriteMovies, onFetchFavouriteMovies }: PropsFromRedux): JSX.Element {
  const [activeCardId, setActiveMovieCardId] = useState<number | undefined>();

  useEffect(() => {
    onFetchFavouriteMovies();
  }, [onFetchFavouriteMovies]);


  const renderMyMovie = (favouriteSingleCard: IMovie): React.ReactNode => (
    <MovieCardElement
      key={favouriteSingleCard.id}
      activeCardId={activeCardId}
      filmName={favouriteSingleCard.name}
      videolink={favouriteSingleCard.poster_image}
      posterSrc={favouriteSingleCard.poster_image}
      id={favouriteSingleCard.id}
      onMouseOver={setActiveMovieCardId}
    />);

  return (
    <div className="user-page">
      <header className="page-header user-page__head">
        <HeaderLogo />
        <h1 className="page-title user-page__title">My list</h1>
        <HeaderUserBlock />
      </header>

      <section className="catalog">
        <h2 className="catalog__title visually-hidden">Catalog</h2>

        <div className="catalog__films-list">

          {favouriteMovies.map(renderMyMovie)}
        </div>
      </section>
      <Footer />
    </div>
  );
}

export { MyList };
export default connector(MyList);
