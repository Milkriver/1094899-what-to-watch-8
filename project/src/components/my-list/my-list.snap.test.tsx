import { render } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router';
import { Provider } from 'react-redux';
import { IMovie } from '../../types/common';
import { configureMockStore } from '@jedmao/redux-mock-store';
import { MyList } from './my-list';


const MOCK_MOVIE_1: IMovie = { id: 1, name: '', posterImage: '', previewImage: '', backgroundImage: '', backgroundColor: '', videoLink: '', previewVideoLink: '', description: '', rating: 1, scoresCount: 1, director: '', starring: ['', ''], runTime: 1, genre: '', released: 1, isFavorite: true };
const MOCK_MOVIE_2: IMovie = { id: 2, name: '', posterImage: '', previewImage: '', backgroundImage: '', backgroundColor: '', videoLink: '', previewVideoLink: '', description: '', rating: 2, scoresCount: 2, director: '', starring: ['', ''], runTime: 1, genre: '', released: 2, isFavorite: false };
const MOCK_MOVIES_ARRAY: IMovie[] = [MOCK_MOVIE_2, MOCK_MOVIE_1];
const mockFn = () => Promise.resolve();
describe('Component: MoviesList', () => {
  const mockStore = configureMockStore();
  const store = mockStore({
    MOVIES: {
      movie: {
        id: 1, name: '', posterImage: '', previewImage: '', backgroundImage: '',
        backgroundColor: '', videoLink: '', previewVideoLink: '', description: '', rating: 1, scoresCount: 1, director: '', starring: ['', ''],
        runTime: 1, genre: '', released: 1, isFavorite: true,
      },
    },
    USER: {},
    REVIEWS: {},
  });

  store.dispatch = jest.fn();

  test('should render correctly', () => {
    const history = createMemoryHistory();
    const page = (
      <Provider store={store}>
        <Router history={history}>
          <MyList
            favouriteMovies={MOCK_MOVIES_ARRAY}
            onFetchFavouriteMovies={mockFn}
          />;
        </Router>
      </Provider>
    );
    const { container } = render(page);
    expect(container).toMatchSnapshot();
  });
});
