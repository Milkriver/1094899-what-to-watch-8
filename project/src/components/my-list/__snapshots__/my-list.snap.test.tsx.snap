// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: MoviesList should render correctly 1`] = `
<div>
  <div
    class="user-page"
  >
    <header
      class="page-header user-page__head"
    >
      <div
        class="logo"
      >
        <a
          class="logo__link"
          href="/"
        >
          <span
            class="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            class="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            class="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <h1
        class="page-title user-page__title"
      >
        My list
      </h1>
      <div
        class="user-block"
      >
        <a
          class="user-block__link"
          href="/login"
        >
          Sign in
        </a>
      </div>
    </header>
    <section
      class="catalog"
    >
      <h2
        class="catalog__title visually-hidden"
      >
        Catalog
      </h2>
      <div
        class="catalog__films-list"
      >
        <article
          class="small-film-card catalog__films-card"
        >
          <div
            class="small-film-card__image"
          >
            <img
              alt=""
              height="175"
              src="posterImage2"
              width="280"
            />
            ;
          </div>
          <h3
            class="small-film-card__title"
          >
            <div
              class="small-film-card__link"
            >
              name2
            </div>
          </h3>
        </article>
        <article
          class="small-film-card catalog__films-card"
        >
          <div
            class="small-film-card__image"
          >
            <img
              alt=""
              height="175"
              src="posterImage1"
              width="280"
            />
            ;
          </div>
          <h3
            class="small-film-card__title"
          >
            <div
              class="small-film-card__link"
            >
              name1
            </div>
          </h3>
        </article>
      </div>
    </section>
    <footer
      class="page-footer"
    >
      <div
        class="logo"
      >
        <a
          class="logo__link logo__link--light"
          href="/"
        >
          <span
            class="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            class="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            class="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        class="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>
</div>
`;
